classes/submission/Repository.php

```diff
@@ -514,18 +514,34 @@ public function canCurrentUserDelete(Submission $submission): bool
      */
     public function canEditPublication(int $submissionId, int $userId): bool
     {
-        // Replaces StageAssignmentDAO::getBySubmissionAndUserIdAndStageId
-        $stageAssignments = StageAssignment::withSubmissionIds([$submissionId])
+        // block authors can never edit a published publication even if an editor granted them canChangeMetadata
+        $assignments = StageAssignment::withSubmissionIds([$submissionId])
             ->withUserId($userId)
             ->get();
 
-        // Check for permission from stage assignments
-        if ($stageAssignments->contains(fn ($stageAssignment) => $stageAssignment->canChangeMetadata)) {
+        $submission = $this->get($submissionId);
+        if ($submission) {
+            foreach ($submission->getData('publications') as $pub) {
+                $status = $pub->getData('status');
+                // if any publication is published or scheduled
+                if (in_array($status, [Submission::STATUS_PUBLISHED, Submission::STATUS_SCHEDULED])) {
+                    $hasNonAuthor = $assignments->contains(fn($sa) =>
```

Comment:
Such checks are better to be done outside of the loop, to avoid executing several times per each publication.

Reply:
Also, we are including `ROLE_ID_ASSISTANT` here but in several places, e.g., preprint grid handler, catalog grid handler, we don't. Is it done deliberately?